### PR TITLE
Check passwords against the HIBP online service

### LINKED ISSUE #1083: Check passwords against hacked password databases [$300]
haveibeenpwned.com has an API to check if passwords have been released in a previous dump.
They also provide a sha1-hashed download of their password dump at https://haveibeenpwned.com/Passwords .
Their API guideline is available here: https://haveibeenpwned.com/API/v2

## Expected Behavior
Ability to check all or a selection of passwords against a password dump (either **downloaded** or online)

## Current Behavior
Currently, KeepassXC does not check against a list of passwords that has been previously compromised.

## Possible Solution (bash)
```
#!/usr/bin/env bash
PASSWORD=('p@55w0rd' 'P@55w0rd' 'password')
ONLINE=false
RELDIR="../pwnedPasswords"
OFFLINE_DATABASE=("${RELDIR}/pwned-passwords-1.0.txt" "${RELDIR}/pwned-passwords-update-1.txt" "${RELDIR}/pwned-passwords-update-2.txt")
check() {
	if ${ONLINE} ; then
		RESPONSE_POST=$(curl -s -o /dev/null -w "%{http_code}" --data "Password=${1}" "https://haveibeenpwned.com/api/v2/pwnedpassword")
		sleep 1.6
		RESPONSE_GET=$(curl -s -o /dev/null -w "%{http_code}" "https://haveibeenpwned.com/api/v2/pwnedpassword/${1}")
		RESPONSE=${RESPONSE_GET}
		sleep 1.6
		if [ ${RESPONSE} -eq 200 ]; then 
			echo "Bad password: ${1} (GET: ${RESPONSE_GET}, POST: ${RESPONSE_POST})"
		elif [ ${RESPONSE} -eq 404 ] ; then 
			echo "Good password: ${1} (GET: ${RESPONSE_GET}, POST: ${RESPONSE_POST})"
		elif [ ${RESPONSE} -eq 429 ] ; then
			echo "Too fast (wait >1.5s)"
		fi
	else
		if ! ${2} ; then
			CHECK=($(echo -n ${i} | sha1sum))
		else
			CHECK="${1}"
		fi
		LINES=$(cat ${OFFLINE_DATABASE[@]} | grep -i ${CHECK} | wc -l)
		if [ ${LINES} -gt 0 ] ; then
			echo "Bad password: ${1}"
		else
			echo "Good password: ${1}"
		fi
	fi
}

for i in ${PASSWORD[@]}; do 
	PASSWORDHASH=($(echo -n ${i} | sha1sum))
	check ${i} false
	check ${PASSWORDHASH[0]} true
done 
```
Note: Above script checks offline for security (slow). It can be modified to check online for testing.

## Context
I would like to check for passwords that have been compromised in a breach or set of breaches.
I can currently do this manually with a modified version of the above script, but it is cumbersome and not as secure.

EDIT:
```echo ${i} -> echo -n ${i}```
Add note to script that it is meant for proof of concept and can be modified for offline use.
Modify script to check offline

### PR DESCRIPTION
**Pull request by wolframroesler**
[TIP]:  # ( Provide a general summary of your changes in the title above ^^ )

## Type of change
[NOTE]: # ( Please remove all lines which don't apply. )
- ✅ New feature (non-breaking change which adds functionality)

## Description and Context
[NOTE]: # ( Describe your changes in detail, why is this change required? )
[NOTE]: # ( Describe the context of your change. Explain large code modifications. )
[NOTE]: # ( If it fixes an open issue, please add "Fixes #XXX" as necessary )

Add "HIBP" to the "Database -> Reports" panel to check all passwords in the database against the HIBP online service, and display the findings in a table. The "worst" passwords (found the most times in the HIBP database) are listed at the beginning of the table (just as in Health Check).

Offline checks (against a previously downloaded HIBP database file) will be added in another PR.

Lots of temporary commits, feel free to squash or  fixup.

Fixes #1083 (together with the upcoming PR for offline checks).

## Screenshots
[TIP]:  # ( Do not include screenshots of your actual database! )

Screenshot from 2020-03-14 12-54-55

## Testing strategy
[NOTE]: # ( Please describe in detail how you tested your changes. )
[TIP]:  # ( We expect new code to be covered by unit tests and documented with doc blocks! )

Tested manually.

## Checklist:
[NOTE]: # ( Please go over all the following points. )
[NOTE]: # ( Again, remove any lines which don't apply. )
[NOTE]: # ( Pull Requests that don't fulfill all [REQUIRED] requisites are likely )
[NOTE]: # ( to be sent back to you for correction or will be rejected.  )
- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**

### CODE REVIEW

**Comment by droidmonkey** [comment]
I squashed and rebased all the commits. Only 1 visual comment here:

1. This dialog is way too wordy and steals focus. I would much prefer this to be in the actual HIBP report widget. For example, before permission is granted, a similar message to this is displayed instead of the results table with only one button "Perform Password Analysis". The technical details written in this dialog are irrelevant to the vast majority of users. Those should be hidden behind a "learn more" link that reveals them.
image

Otherwise this performs REALLY REALLY well! I got great results from my personal database. The only issue I saw was my credit card entries where I put my CVC code in the password field. That is fine because it was obvious to me and entry templates will help with that problem in the future.

**Comment by droidmonkey** [comment]
I am checking the code right now, and where it does work there are many efficiencies that can be had. Why do you jump from QString to StdString to C_Str? All operations can be smoothly done in QString. Additionally, you only use the HIBPDownloader once per password then reconstruct it, that is wasteful and the network manager can be loaded up with requests and it will do 5 at a time in batches. This will speed up lookups significantly.

File: src/core/HibpDownloader.cpp [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+     *
+     * Returns the number of times the password is found in breaches, or
+     * 0 if the password is not in the HIBP result.
+     */
+    int pwnCount(const QString& password, const QString& hibpResult)
+    {
+        const auto hash = sha1Hex(password).toStdString();
+        const auto result = hibpResult.toStdString();
+        const auto pHash = hash.c_str();
+        const auto pResult = result.c_str();
```
Comment (droidmonkey) [inline src/core/HibpDownloader.cpp]: You should use standard QString functions to handle all this

File: src/core/HibpDownloader.cpp [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+        } else {
+
+            // Not found
+            return 0;
+        }
+    }
+} // namespace
+
+HibpDownloader::HibpDownloader(QString password, QObject* parent)
```
Comment (droidmonkey) [inline src/core/HibpDownloader.cpp]: Instead of just passing in 1 password, you should pass in a QStringList of passwords and shove them all into the network manager. It will batch process them and fire off a signal when each is completed. You combine each reply with the password that made it to do the post processing.

Look at the custom icon batch downloader code for guidance.
https://github.com/keepassxreboot/keepassxc/blob/b045160e4fcd612def4c5cff55f9469baf12a738/src/gui/IconDownloaderDialog.cpp#L62-L103

File: src/gui/reports/ReportsDialog.cpp [hunk 3]
```diff
@@ -113,16 +120,33 @@ void ReportsDialog::reject()
     emit editFinished(true);
 }
 
-void ReportsDialog::entryActivationSignalReceived(const Group* group, Entry* entry)
+void ReportsDialog::entryActivationSignalReceived(QWidget* sender, const Group* group, Entry* entry)
 {
+    m_sender = sender;
```
Comment (droidmonkey) [inline src/gui/reports/ReportsDialog.cpp]: This is what `Object::sender()` is for. You should not be passing the sender in the signal.

File: src/gui/reports/ReportsWidgetHibp.cpp [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+                if (!entry->isRecycled()) {
+                    const auto found = m_pwPwned.find(entry->password());
+                    if (found != m_pwPwned.end()) {
+                        items.emplace_back(entry, group, *found);
+                    }
+                }
+            }
+        }
+    }
+
```
Comment (droidmonkey) [inline src/gui/reports/ReportsWidgetHibp.cpp]: You should be using standard Qt classes (QList, QPair). You do not need to store the group because you can access that from the entry: `entry->group()`.

File: src/gui/reports/ReportsWidgetHibp.cpp [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+                              : count <= 10 ? tr("10 times")
+                                            : count <= 100 ? tr("100 times")
+                                                           : count <= 1000 ? tr("1000 times")
+                                                                           : count <= 10 * 1000
+                                                                                 ? tr("10,000 times")
+                                                                                 : count <= 100 * 1000
+                                                                                       ? tr("100,000 times")
+                                                                                       : count <= 1000 * 1000
+                                                                                             ? tr("a million times")
+                                                                                             : tr("millions of times"));
```
Comment (droidmonkey) [inline src/gui/reports/ReportsWidgetHibp.cpp]: This is crazy hard to read. Should be a function with a series of if statements instead.

File: src/gui/reports/ReportsWidgetHibp.cpp [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+    // initial pass through the password queue is done, i. e.
+    // when re-validating a modified password after the user
+    // editied an entry. We want that re-validation to happen
+    // invisibly, popping the progress bar back into view would
+    // just be visually disturbing.
+
+    // Handle the next password
+    m_downloader.reset(new HibpDownloader(*m_pwQueue.begin()));
+    connect(&*m_downloader, SIGNAL(finished(const QString&, int)), this, SLOT(checkFinished(const QString&, int)));
+    connect(&*m_downloader,
```
Comment (droidmonkey) [inline src/gui/reports/ReportsWidgetHibp.cpp]: Never do this: `&*object`

File: src/gui/reports/ReportsWidgetHibp.cpp (hunk 5 above)
Comment (wolframroesler) [reply src/gui/reports/ReportsWidgetHibp.cpp]: Looked a lot better before "make format" messed it up. Will turn it into a function.

File: src/gui/reports/ReportsWidgetHibp.cpp (hunk 4 above)
Comment (wolframroesler) [reply src/gui/reports/ReportsWidgetHibp.cpp]: TIL, tx, that's gonna save quite some work.

**Comment by wolframroesler** [comment]
Reworked as requested, please re-review.

One more thing. I tried to put the following at the beginning of HibpDownloader.hpp:

```cpp
#ifndef WITH_XC_NETWORKING
#error This file requires KeePassXC to be built with network support.
#endif
```

to make sure it was included only when KeePassXC was built with network support. To my surprise, the #error fires *all the time*, even when building with `cmake -DWITH_XC_ALL=ON`. In the .cpp files, `#ifdef WITH_XC_NETWORKING` works as expected, but in the .h files, `WITH_XC_NETWORKING` never seems to be defined (same in ReportsWidgetHibp.cpp/.h). Any explanation for that?

**Comment by droidmonkey** [comment]
You need to `#include "config-keepassx.h"` to use the predefines.

**Comment by droidmonkey** [comment]
@wolframroesler for your awareness i have made significant code enhancements locally and will be pushing them soon. I optimized as much as I could and standardized to Qt and code base.

**Comment by droidmonkey** [comment]
Pushed my code changes, here are some screenshots:

image

image

File: src/gui/reports/ReportsWidgetHibp.h [hunk 7]
```diff
@@ -46,33 +47,29 @@ class ReportsWidgetHibp : public QWidget
...
     void emitEntryActivated(const QModelIndex&);
-    void checkFinished(const QString&, int);
-    void checkFailed(const QString&, const QString&);
+    void addHibpResult(const QString&, int);
+    void fetchFailed(const QString&);
 
 private:
     void makeHibpTable();
     void startValidation();
+    QString countToText(int count);
```
Comment (wolframroesler) [inline src/gui/reports/ReportsWidgetHibp.h]: Is there a reason why this is a non-static, non-const member function?

**Comment by wolframroesler** [comment]
@droidmonkey Thanks for the rework, learned a lot from it :) I should have guessed that something like toHex exists in an all-bells-and-whistles-included library like Qt.

I particularly like the reworked warning message that appears when you switch to the HIBP pane, it's much better than my modal dialog. It looks like it's made to add a second button, "Perform Offline Analysis", after which you select your downloaded HIBP file in case you don't want to go online. Will implement that on a new branch after this one has been merged.

About the warning message, it says

> your database passwords will be cryptographically hashed and the first five characters sent securely to this service

Will everyone understand that this is referring to the first five characters of the hash and not of the password?

**Comment by droidmonkey** [comment]
Great point!

**Comment by wolframroesler** [comment]
Fixed two details: fetchFailed had a wrong signature (produced a message on stdout), and it failed to build without networking support.

**Comment by droidmonkey** [comment]
I did actually intend to remove the password QString from fetchFailed since it is unused and doesn't really matter. I completed my original change in the last commit and fixed the caution message. This should be good to go.

**Comment by droidmonkey** [comment]
Claim your bounty good sir!

**Comment by wolframroesler** [comment]
Thanks @droidmonkey :) will claim the bounty when offline checks are done too.

**Comment by droidmonkey** [comment]
Don't wait for that, I think there are some time restrictions on this bounty.

**Comment by wolframroesler** [comment]
https://www.bountysource.com/issues/50410294-check-passwords-against-hacked-password-databases says "You will be able to claim this bounty as soon as the original issue is marked as closed" so I suppose #1083 needs to be closed first.

**Comment by droidmonkey** [comment]
It is closed :grimacing:

**Comment by danielporto** [comment]
I'm so looking forward for this! Thanks a lot!!
